### PR TITLE
[BUGFIX] Prevent "division by zero" errors in Rule-Based Profiler calculations when Batch has zero rows

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
Division by zero errors are prevented by adding `EPSILON` (an infinitesimal positive number) to denominator (e.g., total rows, non-null rows).  This calculation comes up in `MapMetricColumnDomainBuilder`, `SimpleDateFormatStringParameterBuilder`, `RegexPatternStringParameterBuilder`, `PartitionParameterBuilder`, and `MeanUnexpectedMapMetricMultiBatchParameterBuilder`.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-761/GREAT-1237
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/rule_based_profiler/parameter_builder/simple_date_format_string_parameter_builder.py [hunk 1]
```diff
@@ -210,7 +211,7 @@ def _build_parameters(
         # Now obtain 1-dimensional vector of values of computed metric (each element corresponds to a Batch ID).
         metric_values = metric_values[:, 0]
 
-        nonnull_count: int = sum(metric_values)
+        nonnull_count: int = sum(metric_values) + NP_EPSILON
```
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/parameter_builder/simple_date_format_string_parameter_builder.py]: This is no longer an `int`, right?

File: great_expectations/rule_based_profiler/parameter_builder/regex_pattern_string_parameter_builder.py [hunk 2]
```diff
@@ -159,7 +160,7 @@ def _build_parameters(
         # Now obtain 1-dimensional vector of values of computed metric (each element corresponds to a Batch ID).
         metric_values = metric_values[:, 0]
 
-        nonnull_count: int = sum(metric_values)
+        nonnull_count: int = sum(metric_values) + NP_EPSILON
```
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/parameter_builder/regex_pattern_string_parameter_builder.py]: This is no longer an `int`, right?

**Review by NathanFarmer — CHANGES_REQUESTED** [review]
Some type hint changes requested. Also, please help me understand how this does not effect the parameters computed by the `ParameterBuilder` in a meaningful way. Shouldn't this break a test somewhere? Are these `float`s coerced back to `int` downstream?

File: great_expectations/rule_based_profiler/parameter_builder/simple_date_format_string_parameter_builder.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/simple_date_format_string_parameter_builder.py]: @NathanFarmer Good catch -- fixed!  Thanks!

File: great_expectations/rule_based_profiler/parameter_builder/regex_pattern_string_parameter_builder.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/regex_pattern_string_parameter_builder.py]: @NathanFarmer Thank you -- fixed!

**Comment by alexsherstinsky** [comment]
> Some type hint changes requested. Also, please help me understand how this does not effect the parameters computed by the `ParameterBuilder` in a meaningful way. Shouldn't this break a test somewhere? Are these `float`s coerced back to `int` downstream?

@NathanFarmer I fixed the type hints mistakes -- good catch, thanks!

The calculations do not bear any effect other than remove the "division by zero error" because when the non-null count is zero, the `EPSILON` is too tiny to matter, while when it is zero, the numerator insures that the same result is produced.

Thanks!

**Review by NathanFarmer — APPROVED** [review]
Makes sense to me!

**Pull request merged**